README.md
# Express-Portfolio
Travis CI build of master branch:  [![Travis CI build of master branch](https://travis-ci.com/julia-thea/julia-thea.github.io.svg?branch=master)](https://travis-ci.com/github/julia-thea/julia-thea.github.io)

**Express-Portfolio** serves my [portfolio](https://juliathea.com) website. It includes a project gallery, a short bio, a downloadable resume, and an email contact form.

## Install

```npm install```

## Contact form setup

Currently, app.js is set up to use Gmail to send mail using [Nodemailer](https://nodemailer.com/about/). To create your own instance of this app, you would need to add your Gmail account and [an application-specific password](https://support.google.com/accounts/answer/185833?hl=en) to the environment:
```
export GMAIL_USER=youremailaddress@gmail.com 
export GMAIL_PASS=YourGoogleAppSpecificPassword
```

Read more about [using Gmail with Nodemailer](https://nodemailer.com/usage/using-gmail/) on the offical site.

You can alternatively place the GMAIL_USER and GMAIL_PASS variables in a .env file, as detailed below.

## Environment variables using .env

The included .env.example file can be used to create a .env file to hold environment variables in an easily accessible place in the project directory.

## Run

npm start to run on port 8000.

# Dependencies
- [Node.js](https://nodejs.org/en/)
- [Express](https://expressjs.com/)
- [dotenv](https://www.npmjs.com/package/dotenv) loads environment variables from the .env file into process.env
- [Nodemailer](https://nodemailer.com/about/) sends email from the contact form



CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [2.0.0] - 2021-03-04
### Added
* ```CHANGELOG.md``` added
* ```chai``` v4.3.3, ```mocha``` v8.3.0

### Changed
* Previous `test` script in `package.json` changed to `test:mocha`.

### Removed
* ```hbs``` - using ```ejs``` instead
* ```express-flash``` - using ```connect-flash``` instead


test/test.js
var supertest = require("supertest");

var server = supertest.agent("http://localhost:8000");

describe("Integration test initial setup", function() {
  it("Should return home page"), function(done) {
    server
      .get('/')
      .expect(200)
      .end(function(err, res) {
        done();
      })
  }
});

public/js/script.js
const spans = document.querySelectorAll('h1 span')

spans.forEach(index => index.addEventListener('mouseover', function(e){

    index.classList.add('animated', 'rubberBand')
} ))

spans.forEach(index => index.addEventListener('mouseout', function(e){

    index.classList.remove('animated', 'rubberBand')
} ))

const htmlBar = document.querySelector('.bar-html')
const cssBar = document.querySelector('.bar-css')
const jsBar = document.querySelector('.bar-javascript')
const reactBar = document.querySelector('.bar-react')


//For the Skill Bars
var t1 = new TimelineLite()

t1.fromTo(htmlBar, .75, {width: `calc(0% - 6px)`}, {width: `calc(90% - 6px)`, ease: Power4.easeout})
    .fromTo(cssBar, .75, {width: `calc(0% - 6px)`}, {width: `calc(95% - 6px)`, ease: Power4.easeout})
    .fromTo(jsBar, .75, {width: `calc(0% - 6px)`}, {width: `calc(75% - 6px)`, ease: Power4.easeout})
    .fromTo(reactBar, .75, {width: `calc(0% - 6px)`}, {width: `calc(70% - 6px)`, ease: Power4.easeout})


const controller = new ScrollMagic.Controller()
const scene = new ScrollMagic.Scene({
    triggerElement: '.skills',
    triggerHook: 0
}).setTween(t1).addTo(controller)


function showRequiredCategory(event) {

    const getId = event.id
    const links = document.querySelectorAll('.work-category button')
    for(i=0; i<links.length; i++){
        if(links[i].hasAttribute('class')){
            links[i].classList.remove('active')
        }
    }

    event.classList.add('active')
    const getCategory = document.querySelector(`.category-${getId}`)
    const categories = document.querySelectorAll('div[class^= "category-"]')

    for(i=0;i<categories.length; i++) {
        if(categories[i].hasAttribute('class')){
            categories[i].classList.remove('showCategory')
            categories[i].classList.add('hideCategory')
        }
    }

    getCategory.classList.remove('hideCategory')
    getCategory.classList.add('showCategory')

}


$(function()
{
    function after_form_submitted(data)
    {
        if(data.result == 'success')
        {
            $('form#reused_form').hide();
            $('#success_message').show();
            $('#error_message').hide();
        }
        else
        {
            $('#error_message').append('<ul></ul>');

            jQuery.each(data.errors,function(key,val)
            {
                $('#error_message ul').append('<li>'+key+':'+val+'</li>');
            });
            $('#success_message').hide();
            $('#error_message').show();

            //reverse the response on the button
            $('button[type="button"]', $form).each(function()
            {
                $btn = $(this);
                label = $btn.prop('orig_label');
                if(label)
                {
                    $btn.prop('type','submit' );
                    $btn.text(label);
                    $btn.prop('orig_label','');
                }
            });

        }//else
    }

	$('#reused_form').submit(function(e)
      {
        e.preventDefault();

        $form = $(this);
        //show some response on the button
        $('button[type="submit"]', $form).each(function()
        {
            $btn = $(this);
            $btn.prop('type','button' );
            $btn.prop('orig_label',$btn.text());
            $btn.text('Sending ...');
        });


                    $.ajax({
                type: "POST",
                url: 'handler.php',
                data: $form.serialize(),
                success: after_form_submitted,
                dataType: 'json'
            });

      });
});

// Function that handles confirmation message on the form submit event -- this is not currently accessed by HTML correctly
function postCall(event){
    event.preventDefault();
    var formData = {}
    var inputData = document.getElementById("contact-form").elements;
    for(var i = 0; i < inputData.length; i++){
      formData[inputData[i].name] = inputData[i].value;
    }

    $.post("/", formData,function(data, status){
      alert(status);
    //   var messageElement = document.getElementById("popup message");
    //   messageElement.append(messageElement)
      });
    // console.log(formData);


  }

src/app.js
var express = require('express');
var ejs = require('ejs');
var nodemailer = require('nodemailer');
var app = express();
var path = require('path');
var bodyParser = require('body-parser');
require('dotenv').config();
var session = require('express-session');
var flash = require('connect-flash');
var cookieParser = require('cookie-parser')

// Setup handlebars engine and views location
app.engine('ejs', require('ejs').__express);
app.set('views', path.join(__dirname, '../views'));
app.set('view engine', 'ejs');

// Setup public directory to serve
app.use(express.static('./public'));

// Middleware
app.use(cookieParser())
app.use(session({
    secret: 'secret',
    cookie: { maxAge: 60000 },
    resave: false,
    saveUninitialized: false
}));
app.use(flash());
app.use(bodyParser.urlencoded({extended: true}));
app.use(express.static(path.join(__dirname, '../public')));

// Routes
app.get('/',function (req, res) {
  console.log("Landing Page")
  res.status(200).render('pages/index.ejs', { message : req.flash('message') })
});

// POST route from contact form
app.post('/', (req, res) => {
  console.log(req.body);
  // Instantiate the SMTP server

  const smtpTrans = nodemailer.createTransport({
    host: 'smtp.gmail.com',
    port: 465,
    secure: true,
    auth: {
      user: process.env.GMAIL_USER,
      pass: process.env.GMAIL_PASS
   }
  })

    // Specify what the email will look like
  const mailOpts = {
    from: 'Your sender info here', // This is ignored by Gmail
    to: process.env.GMAIL_USER,
    subject: 'New message from contact form at juliathea.com',
    text: `${req.body.name} (${req.body.email}) says: ${req.body.message}`
  }

  // Attempt to send the email
  smtpTrans.sendMail(mailOpts, (error, response) => {
    if (error) {
      req.flash('message', 'Please try again later');
      res.send("Please try again later");
    }
    else {
      req.flash('message', 'Submitted successfully');
      res.send("Submitted successfully!");
    }
  })
})

app.listen(process.env.PORT || 8000, function () {
    console.log('Server is up on port 8000');
});